module/scmi_reset/src/mod_scmi_reset.c

```diff
@@ -0,0 +1,447 @@
+    [SCMI_RESET_REQUEST] = sizeof(struct scmi_reset_request_a2p),
+    [SCMI_RESET_NOTIFY] = sizeof(struct scmi_reset_notify_a2p),
+};
+
+static unsigned int count_domains(void)
+{
+    unsigned int i = 0;
+    unsigned int count = 0;
+
+    for (i = 0; i < scmi_reset_ctx.config->agent_count; i++)
+	count += scmi_reset_ctx.config->agent_table[i].device_count;
+
+    return count;
+}
+
+/*
+ * Reset domain management protocol implementation
+ */
+
+static int protocol_version_handler(fwk_id_t service_id,
+                                    const uint32_t *payload)
+{
+    struct scmi_protocol_version_p2a outmsg = {
+        .status = SCMI_SUCCESS,
+        .version = SCMI_PROTOCOL_VERSION_RESET_DOMAIN,
+    };
+
+    scmi_reset_ctx.scmi_api->respond(service_id, &outmsg, sizeof(outmsg));
+
+    return FWK_SUCCESS;
+}
+
+static int protocol_attributes_handler(fwk_id_t service_id,
+                                       const uint32_t *payload)
+{
+    struct scmi_reset_protocol_attributes_p2a outmsg = {
+        .status = SCMI_SUCCESS,
+    };
+
+    outmsg.attributes = count_domains();
+
+    scmi_reset_ctx.scmi_api->respond(service_id, &outmsg, sizeof(outmsg));
+
+    return FWK_SUCCESS;
+}
+
+static int protocol_message_attributes_handler(fwk_id_t service_id,
+                                               const uint32_t *payload)
+{
+    struct scmi_protocol_message_attributes_p2a outmsg = {
+        .status = SCMI_NOT_FOUND,
+    };
+    size_t outmsg_size = sizeof(outmsg.status);
+    struct scmi_protocol_message_attributes_a2p params = { 0 };
+
+    params = *(const struct scmi_protocol_message_attributes_a2p *)payload;
+
+    if ((params.message_id < FWK_ARRAY_SIZE(msg_handler_table)) &&
+        msg_handler_table[params.message_id]) {
+        outmsg.status = SCMI_SUCCESS;
+        outmsg_size = sizeof(outmsg);
+    }
+
+    scmi_reset_ctx.scmi_api->respond(service_id, &outmsg, outmsg_size);
+
+    return FWK_SUCCESS;
+}
+
+/*
+ * Given a service identifier, retrieve a pointer to its agent's
+ * \c mod_scmi_reset_agent structure within the agent table.
+ */
+static int get_agent_entry(fwk_id_t service_id,
+			   const struct mod_scmi_reset_agent **agent)
+{
+    int status = 0;
+    unsigned int agent_id = 0;
+
+    status = scmi_reset_ctx.scmi_api->get_agent_id(service_id, &agent_id);
+    if (status != FWK_SUCCESS)
+        return status;
+
+    if (agent_id >= scmi_reset_ctx.config->agent_count)
+        return FWK_E_PARAM;
+
+    *agent = &scmi_reset_ctx.agent_table[agent_id];
+
+    return FWK_SUCCESS;
+}
+
+static int get_reset_device(fwk_id_t service_id, unsigned int domain_id,
+                          const struct mod_scmi_reset_device **reset_device)
+{
+    int status = 0;
+    const struct mod_scmi_reset_agent *agent_entry = NULL;
+
+    status = get_agent_entry(service_id, &agent_entry);
+    if (status != FWK_SUCCESS)
+        return status;
+
+    if (domain_id >= agent_entry->device_count)
```

Comment:
You are assuming here that for each device dom_id is linear. What if we have global domain ids, e.,g, 1,2,3...10. and agent1 get 1, 3, 5 and agent2 gets 2,3,4,6,7. 
We need to loop through the domain table for the agent IMHO.